src/rulesets/Base/base.jl

```diff
@@ -91,6 +92,12 @@
 @scalar_rule(fma(x, y, z), (y, x, One()))
 @scalar_rule(max(x, y), @setup(gt = x > y), (gt, !gt))
 @scalar_rule(min(x, y), @setup(gt = x > y), (!gt, gt))
+@scalar_rule(clamp(x, low, high),
+@setup(
+islow = x < low,
+ishigh = high < x),
+(!( islow | ishigh), islow, ishigh)
+)
```

Comment:
We generally try to follow https://github.com/invenia/BlueStyle here, so I would format this more like this:
```suggestion
@scalar_rule(
    clamp(x, low, high),
    @setup(
        islow = x < low
        ishigh = high < x
    ),
    (!(islow | ishigh), islow, ishigh)
)
```
I think, we should also use `Zero()` and `One()` here instead of `Bool`s, which you can write like `islow ? One() : Zero()`. This has the benefit, that it allows us to avoid allocations when multiplying large arrays, for example.

Reply:
This may have similar issues to #187, although (1) it's less obvious to me how often this is broadcasted vs. used in scalar code, and (2) perhaps getting AbstractDifferential here is better than getting Any:
```julia
julia> grad_clamp1(x, lo, hi) = lo<x<hi 

julia> grad_clamp2(x, lo, hi) = lo<x<hi ? One() : Zero()

julia> grad_clamp1.(randn(1,10),0,1)
1×10 BitArray{2}:
 0  0  0  1  1  0  0  0  0  1

julia> grad_clamp2.(randn(1,10),0,1)
1×10 Array{ChainRulesCore.AbstractDifferential,2}:
 One()  One()  One()  One()  One()  Zero()  Zero()  One()  Zero()  Zero()
```